File: centrifuge.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <iomanip>
#include <thread>
#include <mutex>
#include <map>
#include <chrono>

// Mutex for thread-safe access to the map
std::mutex map_mutex;
std::map<int, std::tuple<int, int, double>> results_map;

// Function to check if the centrifuge is balanced
bool balanced(const std::vector<int>& centrifuge) {
    int n = centrifuge.size();
    std::vector<double> mapped;

    // Map positions to angles
    for (int t = 0; t < n; ++t) {
        if (centrifuge[t] == 0) continue; // Skip if no mass
        mapped.push_back(centrifuge[t] * 2 * M_PI * (t + 1) / n);
    }

    // Calculate the center of mass
    double x = 0.0, y = 0.0;
    for (double angle : mapped) {
        x += cos(angle);
        y += sin(angle);
    }

    // Calculate the magnitude of the center of mass
    double com_of_all_tt = round(hypot(x, y) * 10) / 10;

    // Check if the center of mass is at (0, 0)
    return com_of_all_tt == 0.0;
}

// Function to generate combinations and calculate balanced percentage for a given size
void processSize(int size) {
    int totalCombinations = 1 << size; // 2^size combinations
    int count = 0;

    for (int i = 0; i < totalCombinations; ++i) {
        std::vector<int> combo;
        for (int j = 0; j < size; ++j) {
            combo.push_back((i & (1 << j)) ? 1 : 0);
        }

        if (balanced(combo)) {
            count++;
        }
    }

    // Calculate percentage of balanced combinations
    double percentage = (count / static_cast<double>(totalCombinations)) * 100;

    // Store the results in the map
    {
        std::lock_guard<std::mutex> lock(map_mutex);
        results_map[size] = {count, totalCombinations, percentage};
    }
}

int main() {
    const int max_length = 26; // Maximum size to process
    std::vector<std::thread> threads;

    auto start_time = std::chrono::high_resolution_clock::now();

    // Launch threads for each size
    for (int size = 2; size <= max_length; ++size) {
        threads.emplace_back(processSize, size);
    }

    // Wait for all threads to finish
    for (auto& thread : threads) {
        thread.join();
    }

    auto end_time = std::chrono::high_resolution_clock::now();
    double duration = std::chrono::duration_cast<std::chrono::milliseconds>(end_time - start_time).count() / 1000.0;

    // Print the stored results
    for (const auto& [size, data] : results_map) {
        auto [balanced_count, total_count, percentage] = data;
        std::cout << size << " => " << balanced_count << "/" << total_count 
                  << " => " << std::fixed << std::setprecision(3) << percentage << "%\n";
    }

    std::cout << "Time taken: " << duration << " seconds\n";
    return 0;
}




/**
 * @brief 
 * 
g++ -O3 -std=c++17 -pthread centrifuge.cpp -o centrifuge && ./centrifuge
 * 
 */

File: centrifuge.py
import math
from itertools import product
from multiprocessing import Pool


# Function to check if the centrifuge is balanced
def balanced(centrifuge):
    n = len(centrifuge)
    TWO_PI = 2 * math.pi

    x, y = 0.0, 0.0
    for t, v in enumerate(centrifuge):
        if v == 0:  # Skip if no mass
            continue
        angle = v * TWO_PI * (t + 1) / n
        x += math.cos(angle)
        y += math.sin(angle)

    # Calculate the magnitude of the center of mass
    return round(math.hypot(x, y) * 10) / 10 == 0.0


# Function to process a single size
def process_size(val):
    total_combinations = 0
    balanced_count = 0

    for combo in product([0, 1], repeat=val):
        total_combinations += 1
        if balanced(combo):
            balanced_count += 1

    avg = (balanced_count / total_combinations) * 100
    return val, balanced_count, total_combinations, avg


# Main function
if __name__ == "__main__":
    import time
    start_time = time.time()

    rng = range(2, 20)  # Range of sizes to process
    with Pool(processes=4) as pool:
        results = pool.map(process_size, rng)

    for val, balanced_count, total_combinations, avg in results:
        print(f"{val} => {balanced_count}/{total_combinations} => {avg:.3f}%")

    print(f"Time taken: {time.time() - start_time:.2f} seconds")
